server/remote_cache/action_cache_server/action_cache_server.go

```diff
@@ -359,3 +370,35 @@ func (s *ActionCacheServer) maybeInlineOutputFiles(ctx context.Context, req *rep
 	}
 	return nil
 }
+
+// validateRestrictedAccess checks to see if the instance name has a restricted prefix.
+// If it does, validateRestrictedAccess uses the ClientIdentityService to assert that the
+// request comes from a trusted client: the app or an executor.
+// If the client is not trusted, the ClientIdentityService is not available, or there are any other errors,
+// validateRestrictedAccess returns an UnauthenticatedError.
+func (s *ActionCacheServer) validateRestrictedAccess(ctx context.Context, instanceName string) error {
+	if !isRestricted(instanceName) {
+		return nil
+	}
+	if s.env.GetClientIdentityService() == nil {
+		return status.UnauthenticatedError("No client ID service available to check restricted instance name prefix")
+	}
+	identity, err := s.env.GetClientIdentityService().IdentityFromContext(ctx)
+	if err != nil {
+		return status.UnauthenticatedErrorf("Could not check identity for restricted instance name prefix: %s", err)
+	}
+	if identity.Client != interfaces.ClientIdentityApp && identity.Client != interfaces.ClientIdentityExecutor {
+		return status.UnauthenticatedError("Cannot write restricted ActionResult from untrusted client")
```

Comment:
this message is a bit outdated now since it's used for both the write and read path